### DISCUSSION TITLE
Scaled Float Type

### DISCUSSION THREAD

**Discussion by Hvass-Labs** [pydantic/pydantic#4109] [Feature Request]
Hello, I am new to pydantic and it looks like a very useful and cleverly made piece of software! I am trying to get it working with my own data-types and structures, some of which may be useful as standard features in pydantic.

I need to convert strings with post-fix letters into scaled floats, so that the letter k for kilo scales to 1_000, m for million scales to 1_000_000, etc. For example, the string '123.45k' would become the float 123_450, and the string '23.4m' would become the float 23_400_000.

The following is the function I normally use for this. You may snicker at the high level of detail in the commenting, but the idea is that everyone can just read the English comments and understand very quickly what the code is doing. They don't have to be expert Python programmers to understand it. In contrast, I have taken a look at the code-base for pydantic and there are practically no comments - most functions don't even have doc-strings - so it takes immense time and effort for outsiders to try and understand the pydantic code-base.

    # All the possible float scales.
    # %=percent, k=kilo, m=million, b=billion, t=trillion.
    all_float_scales = {'%': 0.01, 'k': 1e3, 'm': 1e6, 'b': 1e9, 't': 1e12}

    def float_scaled(s):
        """
        Convert argument to a float, and if it is a string then the last character
        can optionally be one of the following (case-insensitive):
        
        - '%' for percent (factor 0.01)
        - 'k' for kilo (1e3)
        - 'm' for million (1e6)
        - 'b' for billion (1e9)
        - 't' for trillion (1e12)
        
        :param s: String or another data-type that can be converted to float.
        :return: Float.
        """
        # Is the argument a string?
        if isinstance(s, str):
            # Is the string non-empty?
            if len(s) > 0:
                # Last character of the string.
                last_char = s[-1].lower()
                
                try:
                    # Lookup the scale from the last character.
                    scale = all_float_scales[last_char]

                    # Convert first part of string to scaled float.
                    f = float(s[:-1]) * scale
                except:
                    # Something went wrong. Maybe a scale wasn't used.
                    # So just use standard conversion to float.
                    f = float(s)
            else:
                # The argument string is empty, so raise an error.
                msg = 'Arg \'s\' must be a non-empty string or number.'
                raise TypeError(msg)
        else:
            # For all other data-types, use standard conversion.
            f = float(s)

        return f

The [pydantic docs](https://pydantic-docs.helpmanual.io/usage/types/#custom-data-types) say there are different ways of making custom types in pydantic, but then only show one way of doing it, so I have based my solution on that.

    class FloatScaled(str):
        @classmethod
        def __get_validators__(cls):
            yield cls.validate

        @classmethod
        def __modify_schema__(cls, field_schema):
            # We should set a matching regex-pattern here and give examples.
            # But it is not clear from the docs what the purpose of this is?
            pass

        @classmethod
        def validate(cls, v):
            return float_scaled(s=v)

        def __repr__(self):
            return f'FloatScaled({super().__repr__()})'

We can then use the new data-type in a pydantic data-model:

    from pydantic import BaseModel

    class Foo(BaseModel):
        bar : FloatScaled

    Foo(bar = '123.45k')
    # Result: Foo(bar=123450.0)

    Foo(bar = '23.4m')
    # Result: Foo(bar=23400000.0)
    
    Foo(bar = '34.1%')
    # Result: Foo(bar=0.341)    

In addition to this, it would be nice if we could restrict the allowed scales. For example, it may be useful to only allow percentages / fractions to be written as either the string '0.341' or '34.1%', but disallow the other scales such as kilo, million, etc.

I think this could be useful for other people. But as I am new to pydantic and I don't really understand it very well, I think it might be better if someone more experienced would add this feature, so it gets done correctly. It might also serve as a good additional example in the docs on how to make custom data-types.

Thanks!

**Comment by samuelcolvin** [comment]
I think good to implement this as your own custom type.

There are so many potential nuances that I'm inclined against including this in pydantic.

But if there's sufficient demand, I would consider it.

Either way, I would advise using SI units.

**Comment by Hvass-Labs** [comment]
Thanks for the quick answer!

I see your point, because the letters I listed are common in finance but the SI units are different.

Perhaps the user could supply the mapping to the type `ScaledFloat` and determine whether it should be as pre- or post-fix. This could then be specialized to `FinanceFloat` and another named `SIFloat`.

Let's see if more people need this and have a good solution.

But it might still be a useful example for your docs / tutorials on how to write a custom type in Pydantic, because I found the current example a bit contrived.
